src/main/java/redis/clients/jedis/JedisClusterInfoCache.java

```diff
@@ -79,36 +80,41 @@ public void discoverClusterNodesAndSlots(Jedis jedis) {
   }
 
   public void discoverClusterSlots(Jedis jedis) {
-    w.lock();
+    //If rediscovering is already in process - no need to start one more same rediscovering, just return
+    if (!rediscovering) {
+      w.lock();
+      rediscovering = true;
```

Comment:
Late to the party, but a possible perf enhancement here might be to check `rediscovering` again:

```
if (!rediscovering) {
  w.lock();
  if (rediscovering) {
    return;
  }
```

That way, if multiple threads make it past the initial check when under high load and block at lock acquisition, only the first one will do the work.